File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/tv/season/TvSeasonService.java
package dev.reuss.tmdb.domain.tv.season;

import dev.reuss.tmdb.common.external.ExternalIds;
import dev.reuss.tmdb.common.image.TvSeasonImages;
import dev.reuss.tmdb.domain.tv.season.model.*;
import dev.reuss.tmdb.domain.tv.season.query.TvSeasonVideosQuery;
import dev.reuss.tmdb.query.AppendToResponse;
import dev.reuss.tmdb.query.ChangesQuery;
import dev.reuss.tmdb.query.ImageQuery;
import dev.reuss.tmdb.value.id.TvSeasonId;
import dev.reuss.tmdb.value.id.TvSeasonNumber;
import dev.reuss.tmdb.value.id.TvShowId;
import dev.reuss.tmdb.value.language.Language;

/**
 * Service for TMDB TV season endpoints.
 */
public interface TvSeasonService {

    /**
     * Loads details for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season details
     */
    TvSeasonDetails details(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads details for a TV season using a specific language.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @param language     response language
     * @return localized TV season details
     */
    TvSeasonDetails details(TvShowId seriesId, TvSeasonNumber seasonNumber, Language language);

    /**
     * Loads details for a TV season with appended responses.
     *
     * @param seriesId         the TV series id
     * @param seasonNumber     the season number
     * @param appendToResponse appended responses
     * @return TV season details with appended responses
     */
    TvSeasonDetails details(
            TvShowId seriesId,
            TvSeasonNumber seasonNumber,
            AppendToResponse<TvSeasonAppend> appendToResponse
    );

    /**
     * Loads details for a TV season using a specific language and appended responses.
     *
     * @param seriesId         the TV series id
     * @param seasonNumber     the season number
     * @param language         response language
     * @param appendToResponse appended responses
     * @return localized TV season details with appended responses
     */
    TvSeasonDetails details(
            TvShowId seriesId,
            TvSeasonNumber seasonNumber,
            Language language,
            AppendToResponse<TvSeasonAppend> appendToResponse
    );

    /**
     * Loads aggregate credits for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season aggregate credits
     */
    TvSeasonAggregateCredits aggregateCredits(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads aggregate credits for a TV season using a specific language.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @param language     response language
     * @return localized TV season aggregate credits
     */
    TvSeasonAggregateCredits aggregateCredits(
            TvShowId seriesId,
            TvSeasonNumber seasonNumber,
            Language language
    );

    /**
     * Loads recent changes for a TV season.
     *
     * @param seasonId the TV season id
     * @return TV season changes
     */
    TvSeasonChanges changes(TvSeasonId seasonId);

    /**
     * Loads changes for a TV season using query parameters.
     *
     * @param seasonId the TV season id
     * @param query    changes query
     * @return TV season changes
     */
    TvSeasonChanges changes(TvSeasonId seasonId, ChangesQuery query);

    /**
     * Loads credits for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season credits
     */
    TvSeasonCredits credits(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads credits for a TV season using a specific language.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @param language     response language
     * @return localized TV season credits
     */
    TvSeasonCredits credits(
            TvShowId seriesId,
            TvSeasonNumber seasonNumber,
            Language language
    );

    /**
     * Loads external IDs for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season external IDs
     */
    ExternalIds externalIds(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads translations for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season translations
     */
    TvSeasonTranslations translations(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads videos for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season videos
     */
    TvSeasonVideos videos(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads videos for a TV season using a specific language.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @param language     response language
     * @return localized TV season videos
     */
    TvSeasonVideos videos(TvShowId seriesId, TvSeasonNumber seasonNumber, Language language);

    /**
     * Loads videos for a TV season using query parameters.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @param query        videos query
     * @return TV season videos
     */
    TvSeasonVideos videos(
            TvShowId seriesId,
            TvSeasonNumber seasonNumber,
            TvSeasonVideosQuery query
    );

    /**
     * Loads watch provider availabilities for a TV season.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @return TV season watch providers
     */
    TvSeasonWatchProviders watchProviders(TvShowId seriesId, TvSeasonNumber seasonNumber);

    /**
     * Loads watch provider availabilities for a TV season using a specific language.
     *
     * @param seriesId     the TV series id
     * @param seasonNumber the season number
     * @param language     response language
     * @return localized TV season watch providers
     */
    TvSeasonWatchProviders watchProviders(
            TvShowId seriesId,
            TvSeasonNumber seasonNumber,
            Language language
    );

    /**
     * Loads images for a TV season.
     *
     * @param tvShowId TMDB TV series id
     * @param seasonNumber season number
     * @return TV season images
     */
    TvSeasonImages images(TvShowId tvShowId, int seasonNumber);

    /**
     * Loads images for a TV season using query parameters.
     *
     * @param tvShowId TMDB TV series id
     * @param seasonNumber season number
     * @param query image query
     * @return TV season images
     */
    TvSeasonImages images(TvShowId tvShowId, int seasonNumber, ImageQuery query);

}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/tv/season/TvSeasonService.kt
package dev.reuss.tmdb.domain.tv.season

import dev.reuss.tmdb.common.external.ExternalIds
import dev.reuss.tmdb.common.image.TvSeasonImages
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonAggregateCredits
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonChanges
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonCredits
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonDetails
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonTranslations
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonVideos
import dev.reuss.tmdb.domain.tv.season.model.TvSeasonWatchProviders
import dev.reuss.tmdb.domain.tv.season.query.TvSeasonVideosQuery
import dev.reuss.tmdb.query.AppendToResponse
import dev.reuss.tmdb.query.ChangesQuery
import dev.reuss.tmdb.query.ImageQuery
import dev.reuss.tmdb.value.id.TvSeasonId
import dev.reuss.tmdb.value.id.TvSeasonNumber
import dev.reuss.tmdb.value.id.TvShowId
import dev.reuss.tmdb.value.language.Language

/** Loads details and related metadata for individual TV seasons. */
interface TvSeasonService {
    /** Returns season details using the client's default language. */
    fun details(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonDetails

    /** Returns season details localized in [language]. */
    fun details(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        language: Language,
    ): TvSeasonDetails

    /** Returns season details with the requested sub-resources appended. */
    fun details(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        appendToResponse: AppendToResponse<TvSeasonAppend>,
    ): TvSeasonDetails

    /** Returns localized season details with the requested sub-resources appended. */
    fun details(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        language: Language,
        appendToResponse: AppendToResponse<TvSeasonAppend>,
    ): TvSeasonDetails

    /** Returns credits combined by person, with roles and episode counts. */
    fun aggregateCredits(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonAggregateCredits

    /** Returns credits combined by person and localized in [language]. */
    fun aggregateCredits(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        language: Language,
    ): TvSeasonAggregateCredits

    /** Returns tracked changes for the season identified by its TMDB [seasonId]. */
    fun changes(seasonId: TvSeasonId): TvSeasonChanges

    /** Returns tracked season changes restricted by [query]. */
    fun changes(
        seasonId: TvSeasonId,
        query: ChangesQuery,
    ): TvSeasonChanges

    /** Returns the season's cast and crew credits using the client's default language. */
    fun credits(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonCredits

    /** Returns the season's cast and crew credits localized in [language]. */
    fun credits(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        language: Language,
    ): TvSeasonCredits

    /** Returns identifiers for the season in external databases. */
    fun externalIds(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): ExternalIds

    /** Returns every translation recorded for the season. */
    fun translations(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonTranslations

    /** Returns season videos using the client's default language. */
    fun videos(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonVideos

    /** Returns season videos localized in [language]. */
    fun videos(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        language: Language,
    ): TvSeasonVideos

    /** Returns season videos using the language filters in [query]. */
    fun videos(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        query: TvSeasonVideosQuery,
    ): TvSeasonVideos

    /** Returns regional streaming and purchase offers using the default language. */
    fun watchProviders(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonWatchProviders

    /** Returns regional streaming and purchase offers localized in [language]. */
    fun watchProviders(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        language: Language,
    ): TvSeasonWatchProviders

    /** Returns season posters using the client's default language filters. */
    fun images(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
    ): TvSeasonImages

    /** Returns season posters using the language filters in [query]. */
    fun images(
        seriesId: TvShowId,
        seasonNumber: TvSeasonNumber,
        query: ImageQuery,
    ): TvSeasonImages
}
